lib/experimental/text_layouter.nim

```diff
@@ -0,0 +1,1754 @@
+    ##
+    ## Refer to the corresponding methods of the LytConsole class for
+    ## descriptions of the methods involved.
+    id {.requiresInit.}: int
+    case kind: LayoutElementKind:
+      of lekString:
+        text {.requiresInit.}: LytStrSpan ## Layout element text
+
+      of lekNewlineSpace:
+        spaceNum: int
+
+      of lekLayoutPrint:
+        layout: Layout
+
+      of lekNewline:
+        discard
+
+    indent: bool
+
+  LytEventKind* = enum
+    layEvStr
+    layEvNewline
+    layEvSpaces
+
+  LytEvent* = object
+    case kind*: LytEventKind
+      of layEvStr:
+        str*: LytStr
+
+      of layEvSpaces:
+        spaces*: int
+
+      of layEvNewline:
+        discard
+
+
+  Layout* = ref object
+    ## An object containing a sequence of directives to the console.
+    elements: seq[LayoutElement]
+
+  LytSolution = ref object
+    ## A Solution object effectively maps an integer (the left margin at
+    ## which the solution is placed) to a layout notionally optimal for
+    ## that margin, together with cost information used to evaluate the
+    ## layout. For compactness, the map takes the form of a
+    ## piecewise-linear cost function, with associated layouts.
+    ##
+    ## This object corresponds to a cost function in the main article.
+    ## Layout is constructed from this function.
+
+    id {.requiresInit.}: int
+
+    knots: seq[int] ## a list of ints, specifying the margin settings at
+    ## which the layout changes. Note that the first knot is required to be
+    ## 0.
+    spans: seq[int] ## a list of ints, giving for each knot, the width of
+    ## the corresponding layout in characters.
+    intercepts: seq[float] ## constant cost associated with each knot -
+    ## computed cost of the outputting this layout solution at each know in
+    ## the `knots` list
+    gradients: seq[float] ## at each knot, the rate with which the layout
+    ## cost increases with an additional margin indent of 1 character.
+    layouts: seq[Layout] ## the Layout objects expressing the optimal
+    ## layout between each knot.
+    index: int
+
+  LytBlockKind* = enum
+    bkText ## A layout consisting of a single line of unbroken text.
+    bkLine ## Horizontally stacked lines
+    bkChoice ## Several alternating layouts
+    bkStack ## Vertically stacked layouts
+    bkWrap ## Mulitple blocks wrapped to create lowerst-cost layout
+    bkVerb ## Multiple lines verbatim
+    bkEmpty ## Empty layout block - ignored by `add` etc.
+
+  LytStr* = object
+    ## Single layout string object. It contains all the information
+    ## required to perform the layout and refer back to the original string
+    ## piece if needed.
+    id*: LytStrId ## Id of the original piece of text
+    len*: int ## It's lengh in units (units are specified - can be ASCII or
+              ## unicode or anything else)
+
+  LytStrSpan* = object
+    ## Span of multiple layout strings
+    strs: seq[LytStr]
+    len: int
+
+  LytBlock* = ref object
```

Comment:
do we expect to have have many things refer to the same block?

Reply:
I don't understand the question. A layout block is a single block in the text layout. It does not refer to anything nor does anyone refer to it, aside from parent combinator blocks like line/stack/combinator/choice. If you are asking why the type is a `ref` - because it forms a tree that is supposed to be randomly added to or added into another trees, similar to how AST is constructed.

Reply:
yup, it was because it's a ref.

Reply:
OK, this is a look at separately/later if the value types make sense.